sdk/src/azure/iot/az_iot_pnp_client_twin.c

```diff
@@ -0,0 +1,379 @@
+    {
+      return AZ_ERROR_ITEM_NOT_FOUND;
+    }
+  } while (az_result_succeeded(az_json_reader_next_token(jr)));
+
+  return AZ_ERROR_ITEM_NOT_FOUND;
+}
+
+// Check if the component name is in the model
+static az_result is_component_in_model(
+    az_iot_pnp_client const* client,
+    az_json_token* component_name,
+    az_span* out_component_name)
+{
+  int32_t index = 0;
+
+  while (index < client->_internal.options.component_names_length)
+  {
+    if (az_json_token_is_text_equal(
+            component_name, client->_internal.options.component_names[index]))
+    {
+      *out_component_name = client->_internal.options.component_names[index];
+      return AZ_OK;
+    }
+
+    index++;
+  }
+
+  return AZ_ERROR_UNEXPECTED_CHAR;
+}
+
+AZ_NODISCARD az_result az_iot_pnp_client_twin_get_property_version(
+    az_iot_pnp_client const* client,
+    az_json_reader* json_reader,
+    az_iot_pnp_client_twin_response_type response_type,
+    int32_t* out_version)
+{
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_NOT_NULL(json_reader);
+
+  (void)client;
+
+  az_json_reader copy_json_reader = *json_reader;
+
+  if ((az_result_failed(az_json_reader_next_token(&copy_json_reader)))
+      || (copy_json_reader.token.kind != AZ_JSON_TOKEN_BEGIN_OBJECT)
+      || (az_result_failed(az_json_reader_next_token(&copy_json_reader))))
+  {
+    return AZ_ERROR_UNEXPECTED_CHAR;
+  }
+
+  if (response_type == AZ_IOT_PNP_CLIENT_TWIN_RESPONSE_TYPE_GET
+      && (az_result_failed(json_child_token_move(&copy_json_reader, iot_hub_twin_desired))
+          || (az_result_failed(az_json_reader_next_token(&copy_json_reader)))))
+  {
+    return AZ_ERROR_UNEXPECTED_CHAR;
+  }
+
+  if (az_result_failed(json_child_token_move(&copy_json_reader, iot_hub_twin_desired_version))
+      || az_result_failed(az_json_token_get_int32(&copy_json_reader.token, out_version)))
+  {
+    return AZ_ERROR_UNEXPECTED_CHAR;
+  }
+
+  return AZ_OK;
+}
+
+static az_result check_if_skipable(
+    az_json_reader* jr,
+    az_iot_pnp_client_twin_response_type response_type)
+{
+  // First time move
+  if (jr->_internal.bit_stack._internal.current_depth == 0)
+  {
+    if ((az_result_failed(az_json_reader_next_token(jr)))
+        || (jr->token.kind != AZ_JSON_TOKEN_BEGIN_OBJECT)
+        || (az_result_failed(az_json_reader_next_token(jr))))
+    {
+      return AZ_ERROR_UNEXPECTED_CHAR;
+    }
+
+    if (response_type == AZ_IOT_PNP_CLIENT_TWIN_RESPONSE_TYPE_GET
+        && (az_result_failed(json_child_token_move(jr, iot_hub_twin_desired))
+            || (az_result_failed(az_json_reader_next_token(jr)))))
+    {
+      return AZ_ERROR_UNEXPECTED_CHAR;
+    }
+  }
+  while (1)
+  {
+    if ((response_type == AZ_IOT_PNP_CLIENT_TWIN_RESPONSE_TYPE_DESIRED_PROPERTIES
+         && jr->_internal.bit_stack._internal.current_depth == 1)
+        || (response_type == AZ_IOT_PNP_CLIENT_TWIN_RESPONSE_TYPE_GET
+            && jr->_internal.bit_stack._internal.current_depth == 2))
```

Comment:
Is it guaranteed for what you are looking for to be on these depths?

Reply:
This method could use some inline comments explaining intent.

Reply:
I added a comment with sample JSON documents with labels for the sections these are trying to convey in the latest push